opm/simulators/flow/BlackoilModel.hpp

```diff
@@ -910,12 +912,29 @@ namespace Opm {
             auto cnvErrorPvFraction = computeCnvErrorPv(B_avg, dt);
             cnvErrorPvFraction /= (pvSum - numAquiferPvSum);
 
-            const double tol_mb  = param_.tolerance_mb_;
             // Default value of relaxed_max_pv_fraction_ is 0.03 and min_strict_cnv_iter_ is 0.
             // For each iteration, we need to determine whether to use the relaxed CNV tolerance.
             // To disable the usage of relaxed CNV tolerance, you can set the relaxed_max_pv_fraction_ to be 0.
-            const bool use_relaxed = cnvErrorPvFraction < param_.relaxed_max_pv_fraction_ && iteration >= param_.min_strict_cnv_iter_;
-            const double tol_cnv = use_relaxed ? param_.tolerance_cnv_relaxed_ :  param_.tolerance_cnv_;
+            bool use_relaxed_cnv = cnvErrorPvFraction < param_.relaxed_max_pv_fraction_ && iteration >= param_.min_strict_cnv_iter_;
+            bool use_relaxed_mb = param_.min_strict_mb_iter_ < 0? false : iteration >= param_.min_strict_mb_iter_;
+
+            if  ((iteration + 1) == maxIter) {
+                use_relaxed_mb = true;
+                use_relaxed_cnv = true;
+                if ( terminal_output_ ) {
+                    std::ostringstream ss;
```

Comment:
I suggest using fmt::format instead, that would simplify this as you do not have to worry about restoring flags and precision etc.

Reply:
I have updated accordingly.